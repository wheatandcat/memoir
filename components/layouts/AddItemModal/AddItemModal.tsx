import  { memo, useState, useCallback, useEffect, type FC } from 'react';
import { StyleSheet, TouchableOpacity } from 'react-native';
import View from '@/components/elements/View';
import Modal from '@/components/layouts/Modal';
import TextInput from '@/components/elements/TextInput';
import Categories from '@/components/layouts/Categories';
import Compatibility from '@/components/layouts/Compatibility/Compatibility';
import Text from '@/components/elements/Text';
import dayjs from 'lib/dayjs';
import type { NewItem } from 'queries/api/index';
import usePrevious from 'hooks/usePrevious';
import DateTimePickerModal from 'react-native-modal-datetime-picker';
import theme from 'config/theme';

export type Props = {
  item?: NewItem;
  isVisible: boolean;
  date: string;
  edit?: boolean;
  loading?: boolean;
  onClose: () => void;
  onAdd: (newItem: NewItem) => void;
};

type State = {
  title: string;
  categoryID: number | null;
  date: Date;
  like: boolean;
  dislike: boolean;
};

const initialState = (date: string, item?: NewItem): State => {
  return {
    title: item?.title || '',
    categoryID: item?.categoryID || null,
    date: new Date(date),
    like: item?.like || false,
    dislike: item?.dislike || false,
  };
};

const AddItemModal: FC<Props> = ({
  edit = false,
  loading = false,
  ...props
}) => {
  const [state, setState] = useState<State>(
    initialState(props.date, props.item)
  );
  const prevIsVisible = usePrevious(props.isVisible);
  const [openDate, setOpenDate] = useState(false);

  useEffect(() => {
    if (props.isVisible && props.isVisible !== prevIsVisible) {
      setState(initialState(props.date, props.item));
    }
  }, [props.isVisible, props.item, props.date, prevIsVisible]);

  const onCategory = useCallback((categoryID: number) => {
    setState((s) => ({ ...s, categoryID }));
  }, []);

  const onChangeTitle = useCallback((title: string) => {
    setState((s) => ({ ...s, title }));
  }, []);

  const onChangeDate = useCallback((date: Date) => {
    setState((s) => ({ ...s, date }));
  }, []);

  const onLike = useCallback(() => {
    setState((s) => ({
      ...s,
      like: !s.like,
      dislike: !s.like ? false : s.dislike,
    }));
  }, []);

  const onDislike = useCallback(() => {
    setState((s) => ({
      ...s,
      dislike: !s.dislike,
      like: !s.dislike ? false : s.like,
    }));
  }, []);

  const onAdd = useCallback(() => {
    const item: NewItem = {
      title: state.title,
      categoryID: state.categoryID || 0,
      date: dayjs(state.date).format('YYYY-MM-DDT00:00:00+09:00'),
      like: state.like,
      dislike: state.dislike,
    };

    props.onAdd(item);
  }, [props.onAdd, state]);

  const valid = useCallback(() => {
    if (state.title === '') {
      return false;
    }

    if (state.categoryID === null) {
      return false;
    }

    return true;
  }, [state.title, state.categoryID]);

  return (
    <Modal
      isVisible={props.isVisible}
      title={dayjs(props.date).format('YYYY.MM.DD / ddd')}
      titleElement={
        edit ? (
          <TouchableOpacity onPress={() => setOpenDate(!openDate)}>
            <View>
              <Text variants="middle" textAlign="center">
                {dayjs(state.date).format('YYYY.MM.DD / ddd')}
              </Text>
            </View>
          </TouchableOpacity>
        ) : undefined
      }
      onClose={props.onClose}
      buttonTitle="入力"
      disabledButton={!valid() && !loading}
      onPress={onAdd}
      loading={loading}
      testID="add_item_modal"
    >
      <View style={styles.root} py={2} px={3}>
        <DateTimePickerModal
          isVisible={openDate && !!edit}
          mode="date"
          locale="ja"
          confirmTextIOS="完了"
          cancelTextIOS="キャンセル"
          is24Hour={true}
          date={state.date}
          onConfirm={(v) => {
            setOpenDate(false);
            onChangeDate(v);
          }}
          onCancel={() => setOpenDate(false)}
          onResponderEnd={() => setOpenDate(false)}
        />
        <TextInput
          placeholder="今日何やった？"
          onChangeText={onChangeTitle}
          autoFocus
          returnKeyType="done"
          defaultValue={state.title}
          style={state.title ? styles.input : styles.placeholder}
        />

        <View py={3}>
          <Categories categoryID={state.categoryID} onPress={onCategory} />
        </View>
        <View py={3}>
          <Compatibility
            like={state.like}
            dislike={state.dislike}
            onLike={onLike}
            onDislike={onDislike}
          />
        </View>
      </View>
    </Modal>
  );
};

export default memo(AddItemModal);

const styles = StyleSheet.create({
  root: {},
  placeholder: {
    height: 60,
    fontSize: theme().fontSizes.xl,
  },
  input: {
    height: 60,
    fontSize: theme().fontSizes.lg,
  },
});
