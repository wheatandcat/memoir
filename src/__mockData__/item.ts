import { ItemQuery } from 'queries/api/index';
type Item = NonNullable<ItemQuery['item']>;

export const item = (option?: Partial<Item>): Item => ({
  id: '1',
  title: '買い物',
  categoryID: 9,
  date: '2021-01-01T00:00:00+09:00',
  like: false,
  dislike: false,
  createdAt: '2021-01-01T00:00:00+09:00',
  updatedAt: '2021-01-01T00:00:00+09:00',
  ...option,
});

export const items = (): Item[] => [
  item(),
  item({
    id: '2',
    title: '本を読む',
    categoryID: 2,
    date: '',
    dislike: false,
    like: false,
  }),
];
