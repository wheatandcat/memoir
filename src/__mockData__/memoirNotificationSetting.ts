import { State as MemoirNotificationSettingState } from 'hooks/useMemoirNotificationSetting';

export const memoirNotificationSetting = (
  option?: Partial<MemoirNotificationSettingState>
): MemoirNotificationSettingState => ({
  dayOfWeek: 0,
  hours: 0,
  minutes: 0,
  notification: false,
  ...option,
});
