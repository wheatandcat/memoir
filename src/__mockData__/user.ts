import { User } from 'queries/api/index';

export const user = (option?: Partial<User>): User => ({
  id: '1',
  displayName: 'test',
  image: '',
  updatedAt: '2021-01-01T00:00:00+09:00',
  createdAt: '2021-01-01T00:00:00+09:00',
  ...option,
});

export const users = (): User[] => [
  {
    id: 'test1',
    displayName: 'suzuki1',
    image: 'https://placehold.jp/150x150.png',
    updatedAt: '2021-01-01T00:00:00+09:00',
    createdAt: '2021-01-01T00:00:00+09:00',
  },
  {
    id: 'test2',
    displayName: 'suzuki2',
    image: 'https://placehold.jp/150x150.png',
    updatedAt: '2021-01-01T00:00:00+09:00',
    createdAt: '2021-01-01T00:00:00+09:00',
  },
  {
    id: 'test3',
    displayName: 'suzuki4',
    image: 'https://placehold.jp/150x150.png',
    updatedAt: '2021-01-01T00:00:00+09:00',
    createdAt: '2021-01-01T00:00:00+09:00',
  },
  {
    id: 'test4',
    displayName: 'suzuki4',
    image: 'https://placehold.jp/150x150.png',
    updatedAt: '2021-01-01T00:00:00+09:00',
    createdAt: '2021-01-01T00:00:00+09:00',
  },
  {
    id: 'test5',
    displayName: 'suzuki5',
    image: 'https://placehold.jp/150x150.png',
    updatedAt: '2021-01-01T00:00:00+09:00',
    createdAt: '2021-01-01T00:00:00+09:00',
  },
];
