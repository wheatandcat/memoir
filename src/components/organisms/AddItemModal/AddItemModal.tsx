import React, { memo, useState, useCallback, useEffect } from 'react';
import { StyleSheet } from 'react-native';
import View from 'components/atoms/View';
import Modal from 'components/organisms/Modal';
import TextInput from 'components/atoms/TextInput';
import Categories from 'components/organisms/Categories';
import Compatibility from 'components/organisms/Compatibility/Compatibility';
import dayjs from 'dayjs';
import advancedFormat from 'dayjs/plugin/advancedFormat';
import 'dayjs/locale/ja';
import { NewItem } from 'queries/api/index';
import usePrevious from 'hooks/usePrevious';
import theme from 'config/theme';

dayjs.locale('ja');
dayjs.extend(advancedFormat);

type Props = {
  item?: NewItem;
  isVisible: boolean;
  loading: boolean;
  date: string;
  onClose: () => void;
  onAdd: (newItem: NewItem) => void;
};

type State = {
  title: string;
  categoryID: number | null;
  like: boolean;
  dislike: boolean;
};

const initialState = (item?: NewItem): State => {
  return {
    title: item?.title || '',
    categoryID: item?.categoryID || null,
    like: item?.like || false,
    dislike: item?.dislike || false,
  };
};

const AddItemModal: React.FC<Props> = (props) => {
  const [state, setState] = useState<State>(initialState(props.item));
  const prevIsVisible = usePrevious(props.isVisible);

  useEffect(() => {
    if (props.isVisible && props.isVisible !== prevIsVisible) {
      setState(initialState(props.item));
    }
  }, [props.isVisible, props.item, prevIsVisible]);

  const onCategory = useCallback((categoryID: number) => {
    setState((s) => ({ ...s, categoryID }));
  }, []);

  const onChangeTitle = useCallback((title: string) => {
    setState((s) => ({ ...s, title }));
  }, []);

  const onLike = useCallback(() => {
    setState((s) => ({
      ...s,
      like: !s.like,
      dislike: !s.like ? false : s.dislike,
    }));
  }, []);

  const onDislike = useCallback(() => {
    setState((s) => ({
      ...s,
      dislike: !s.dislike,
      like: !s.dislike ? false : s.like,
    }));
  }, []);

  const onAdd = useCallback(() => {
    const item: NewItem = {
      title: state.title,
      categoryID: state.categoryID || 0,
      date: dayjs(props.date).format('YYYY-MM-DDT00:00:00+09:00'),
      like: state.like,
      dislike: state.dislike,
    };

    props.onAdd(item);
  }, [props, state]);

  const valid = useCallback(() => {
    if (state.title === '') {
      return false;
    }

    if (state.categoryID === null) {
      return false;
    }

    return true;
  }, [state]);

  return (
    <Modal
      isVisible={props.isVisible}
      title={dayjs(props.date).format('YYYY.MM.DD / ddd')}
      onClose={props.onClose}
      buttonTitle="入力"
      disabledButton={!valid() && !props.loading}
      onPress={onAdd}
      loading={props.loading}
    >
      <View style={styles.root} py={2} px={3}>
        <TextInput
          placeholder="今日何やった？"
          onChangeText={onChangeTitle}
          autoFocus
          returnKeyType="done"
          defaultValue={state.title}
          style={state.title ? styles.input : styles.placeholder}
        />

        <View py={3}>
          <Categories categoryID={state.categoryID} onPress={onCategory} />
        </View>
        <View py={3}>
          <Compatibility
            like={state.like}
            dislike={state.dislike}
            onLike={onLike}
            onDislike={onDislike}
          />
        </View>
      </View>
    </Modal>
  );
};

export default memo(AddItemModal);

const styles = StyleSheet.create({
  root: {},
  placeholder: {
    height: 60,
    fontSize: theme().fontSizes.xl,
  },
  input: {
    height: 60,
    fontSize: theme().fontSizes.lg,
  },
});
