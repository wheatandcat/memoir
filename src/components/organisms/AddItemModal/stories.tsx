import React from 'react';
import { storiesOf } from '@storybook/react-native';
import View from 'components/atoms/View';
import { mockFn } from 'storyBookUtils/index';
import AddItemModal from './';

storiesOf('organisms', module).add('AddItemModal', () => (
  <View>
    <AddItemModal isVisible date="2020-01-01" onClose={mockFn('onClose')} />
  </View>
));
