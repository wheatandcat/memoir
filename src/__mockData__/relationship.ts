import type { RelationshipsQuery } from "queries/api/index";
import { user } from "./user";

export type Relationship = NonNullable<
  EdgesNode<RelationshipsQuery["relationships"]>
>;

export const relationship = (option?: Partial<Relationship>): Relationship => ({
  id: "1",
  followedId: "test",
  followerId: "test",
  user: user(),
  createdAt: "2021-01-01T00:00:00+09:00",
  updatedAt: "2021-01-01T00:00:00+09:00",
  ...option,
});

export const relationships = (): Relationship[] => [
  relationship(),
  relationship({
    id: "2",
    user: user(),
  }),
  relationship({
    id: "3",
    user: user(),
  }),
  relationship({
    id: "4",
    user: user(),
  }),
];
