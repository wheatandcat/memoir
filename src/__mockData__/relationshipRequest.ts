import { RelationshipRequestsQuery } from 'queries/api/index';
import { user } from './user';

export type RelationshipRequest = NonNullable<
  EdgesNode<RelationshipRequestsQuery['relationshipRequests']>
>;

export const relationshipRequest = (
  option?: Partial<RelationshipRequest>
): RelationshipRequest => ({
  id: '1',
  followedId: 'test',
  followerId: 'test',
  user: user(),
  createdAt: '2021-01-01T00:00:00+09:00',
  updatedAt: '2021-01-01T00:00:00+09:00',
  ...option,
});

export const relationshipRequests = (): RelationshipRequest[] => [
  relationshipRequest(),
  relationshipRequest({
    id: '2',
    user: user(),
  }),
  relationshipRequest({
    id: '3',
    user: user(),
  }),
  relationshipRequest({
    id: '4',
    user: user(),
  }),
];
