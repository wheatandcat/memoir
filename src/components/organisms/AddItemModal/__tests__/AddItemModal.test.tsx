import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { item } from '__mockData__/item';
import AddItemModal, { Props } from '../';

const propsData = (): Props => ({
  item: {
    categoryID: item().categoryID,
    date: item().date,
    dislike: item().dislike,
    like: item().like,
    title: item().title,
  },
  isVisible: true,
  loading: false,
  date: '2020-01-01',
  onAdd: jest.fn(),
  onClose: jest.fn(),
});

describe('components/organisms/AddItemModal/AddItemModal.tsx', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<AddItemModal {...propsData()} />);
  });

  it('正常にrenderすること', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
