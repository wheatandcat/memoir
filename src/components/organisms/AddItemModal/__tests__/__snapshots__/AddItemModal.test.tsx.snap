// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/organisms/AddItemModal/AddItemModal.tsx 正常にrenderすること 1`] = `
<Modal
  buttonTitle="入力"
  disabledButton={false}
  isVisible={true}
  loading={false}
  onClose={[MockFunction]}
  onPress={[Function]}
  title="2020.01.01 / 水"
>
  <View
    px={3}
    py={2}
    style={Object {}}
  >
    <TextInput
      autoFocus={true}
      defaultValue="買い物"
      onChangeText={[Function]}
      placeholder="今日何やった？"
      returnKeyType="done"
      style={
        Object {
          "fontSize": 24,
          "height": 60,
        }
      }
    />
    <View
      py={3}
    >
      <Memo(Categories)
        categoryID={9}
        onPress={[Function]}
      />
    </View>
    <View
      py={3}
    >
      <Memo(Compatibility)
        dislike={false}
        like={false}
        onDislike={[Function]}
        onLike={[Function]}
      />
    </View>
  </View>
</Modal>
`;
